hilti/toolchain/src/compiler/global-optimizer.cc

```diff
@@ -289,25 +299,35 @@ struct Visitor : hilti::visitor::PreOrder<bool, Visitor> {
 
                 switch ( fn.callingConvention() ) {
                     case function::CallingConvention::ExternNoSuspend:
-                    case function::CallingConvention::Extern:
-                        // If the declaration is `extern` it is part of an externally
-                        // visible API and potentially used elsewhere.
-                        function.referenced = true;
+                    case function::CallingConvention::Extern: {
+                        // If the declaration is `extern` and the unit is `public` or a filter, the function
+                        // is part of an externally visible API and potentially used elsewhere.
```

Comment:
I haven't quite understood yet why filter types are special: HILTI doesn't know anything about them, so I would think that normal visibility rules apply. Where's a filter function used that goes around that?

Reply:
Filters are not in general `public`, so with how this patch is structured their parse functions are candidates for removal if we see no usages. The uses of these functions are typically in code generated for `connect_filter`, but not visible on the AST level.

The right approach is probably to hardcode a whitelist of functions which make use of certain unit member functions (e.g., into the visitor for `MemberCall`), and then explicitly mark the unit member functions as used from there. I'll give that a try.

Reply:
I instead went with the approach discussed offline, namely explicitly marking parse functions of filters as `&always-emit`.